Client server article (#17)
bonsai-rx/zeromq: docs/tutorials/client-server.md

:::workflow
![Multiple clients](~/workflows/multiple-clients.bonsai)
:::

> [!Note]
> For the purposes of this article we are creating all of our clients and our server on the same Bonsai project and same machine for ease of demonstration. In a working example, each client and server could be running in separate Bonsai instances on different machines on a network. In this case, localhost would be replaced with the server machine’s IP address.

## Basic server
Now that we have our client pool set up and sending messages, let’s implement a server to listen for those messages. 
- Add a [`Router`](xref:Bonsai.ZeroMQ.Router) operator to the project and set its properties to match the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) sockets we already added so that it is running on the same network. 
- As the [`Router`](xref:Bonsai.ZeroMQ.Router) is acting as server and will be the ‘static’ part of the network, set its `Action` to ‘Bind’.

> As with the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator, a [`Router`](xref:Bonsai.ZeroMQ.Router) operator without any input will simply listen for messages on the network and not send anything in return. If we run the project now and monitor the output of the [`Router`](xref:Bonsai.ZeroMQ.Router) operator, we'll see that each time the client sends a message triggered by its associated key press we get a `ResponseContext` produced at the [`Router`](xref:Bonsai.ZeroMQ.Router). Expanding the output the the [`Router`](xref:Bonsai.ZeroMQ.Router), we can see it contains a `NetMQMessage`. We [expect](https://netmq.readthedocs.io/en/latest/router-dealer/) this message to be composed of 3 frames: an address (in this case the address of the client that sent the message), an empty delimiter frame and the message content. 

Reviewer: Extra "the" in this sentence: "expanding the output the the".

- Expose the `Buffer` `byte[]` of the `First` frame. 

> - Add an [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator the the first frame buffer and set its `Value` to 1 to access the unique address ID. 

Reviewer: Extra "the" in "the the first frame buffer"

> [!Note]
> Imagine, for example, that our Dealer sockets were sending video snippets to a Router server that is tasked with doing some processing of the video and returning the results back to the Dealers. If the responses were not computed in an asynchronous manner we would start to incur a bottleneck on the router if there were many connected Dealers or frequent Dealer requests.

Running this workflow, you should see a 'bounceback' where any [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) client that sends a message receives a reply from the [`Router`](xref:Bonsai.ZeroMQ.Router) server. However, in order to address these messages to specific other clients we need to take a slightly different approach. 

- Delete the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) and [`ConvertToString`](xref:Bonsai.ZeroMQ.ConvertToString) branches.
- Replace with a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) called `BouceBack` that generates a bounceback message without using the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator: 

:::workflow
![Server message multicast](~/workflows/server-message-multicast.bonsai)
:::

> We had to change quite a few things to modify this workflow so let's step through the general logic. The first thing to note is that since we are avoiding the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator in this implementation we need to pass messages directly into the [`Router`](xref:Bonsai.ZeroMQ.Router). To do this we generate a [`BehaviorSubject`](xref:Bonsai.Reactive.BehaviorSubject) source with a `NetMQMessage` output type and connect it to the [`Router`](xref:Bonsai.ZeroMQ.Response) (can implement this by creating a [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) operator, right-clicking it and creating a [`BehaviorSubject`](xref:Bonsai.Reactive.BehaviorSubject) source). This will change the output type of the [`Router`](xref:Bonsai.ZeroMQ.Router) operator from a `ResponseContext` to a `NetMQMessage` so we need to make some modifications to how we process the stream.

Reviewer: I got stuck here, I couldn't create a `BehaviourSubject` from `ToMessage` unless the `Bounceback` logic was completed since there was no `NetMQMessage` input. Maybe move it to after the next section. Consider a step breakout for this paragraph.

> We want the [`Router`](xref:Bonsai.ZeroMQ.Router) to generate a reply message every time it receives a request from a [`Dealer`](xref:Bonsai.ZeroMQ.Dealer). Since we are now building this message ourselves instead of using [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse), we branch off the [`Router`](xref:Bonsai.ZeroMQ.Router) with a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator. Inside, we split the `NetMQMessage` into its component `NetMQFrame` parts, taking the `First` frame for the address, using [`Index`](xref:Bonsai.Expressions.IndexBuilder) to grab the middle empty delimiter frame and creating a new `String` which we convert to a `NetMQFrame` for the message content. We [`Merge`](xref:Bonsai.Reactive.Merge) these component frames back together and use a [`Take`](xref:Bonsai.Reactive.Take) operator (with count = 3) followed by [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage). The [`Take`](xref:Bonsai.Reactive.Take) operator is particularly important here as 1) [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) will only complete the message once the observable stream is completed and 2) We need to close the observable anyway to complete the [`SelectMany`](xref:Bonsai.Reactive.SelectMany). Finally, we use a [`MulticastSubject`](xref:Bonsai.Expressions.MulticastSubject) to send our completed message to the [`Router`](xref:Bonsai.ZeroMQ.Router).

Reviewer: Consider a step breakout for this paragraph too.

Author: I've split this into a step breakout for the actual construction and left this paragraph as explanation as it's probably the bit that benefits most from more detailed breakdown of the logic.

If we run the workflow now, we should see the same behavior as before (server bounces message back to initiating client).

## SelectMany detour
Now our network has a complete loop of client --> server --> client communication, but only the client that sends a message receives anything back from the server. Instead we’d like all clients to know when any of the clients sends a message. We already have access to the connected clients from `ClientAddresses`, and we know how to package data and send it back to clients via the [`Router`](xref:Bonsai.ZeroMQ.Router). In an imperative language we would do something like:

```
foreach (var client in Clients) {
    Router.SendMessage(client.Address, Message);
}
```

using a loop to send the message back to each client in turn. In a reactive / observable sequence based framework we have to think about this a bit differently. The solution is to use a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator and it is worth taking a detour here to understand its use in some detail since we have already used it to generate our bounceback message and will apply it again for addressing multiple clients.

> The [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator can be a tricky one to understand. Lee Campbell’s excellent [Introduction to Rx](http://introtorx.com/Content/v1.0.10621.0/08_Transformation.html#SelectMany) book does a good job of summarising its utility, suggesting we think of it as “from one, select many” or “from one, select zero or more”. In our case, we can think of [`SelectMany`](xref:Bonsai.Reactive.SelectMany) as a way to repeat some processing logic several times and feed the output of each repeat into a single sequence. More concretely, taking a single message and repeating the act of sending it several times for each client address. It is easier to show by example, so let’s set up a toy example in our project. 

Reviewer: Broken link for `Introduction to Rx` book. I found the updated link here https://introtorx.com/chapters/transformation-of-sequences (I could not find a section anchor link), but I am also not sure how helpful it will be as it is quite dense with Rx terminology which people might not be familiar with. Consider removing the link and making the section more concise.

> Create a [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) operator followed by a [`SelectMany`](xref:Bonsai.Reactive.SelectMany). Set the `Filter` for the [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) to a key that hasn’t been assigned to a client yet – here I will use ‘A’. Inside the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) oeprator add a [`SubscribeSubject`](xref:Bonsai.Expressions.SubscribeSubject) and set its subscription to the `ClientAddresses` subject we created earlier to replay unique client addresses. Add a [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil) operator after the [`SubscribeSubject`](xref:Bonsai.Expressions.SubscribeSubject) and connect the output of [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil) to the [`WorkflowOutput`](xref:Bonsai.Expressions.WorkflowOutputBuilder) (disconnecting the `Source` operator). Finally, create a [`KeyUp`](xref:Bonsai.Windows.Input.KeyUp) operator and connect it to [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil). Set the key `Filter` for [`KeyUp`](xref:Bonsai.Windows.Input.KeyUp) to the same as previously created [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) operator outside the [`SelectMany`](xref:Bonsai.Reactive.SelectMany).

Reviewer: Spelling error- "oeprator". Also consider a step breakout for this section?

:::

The [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator has a couple of interesting properties which may not be immediately obvious from this simple example. First, this operator always transmits its response back to the ZeroMQ socket that initiated the request (in this case one of our [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) clients) and we therefore do not need to specify an address in its processing logic. Second, the internal flow of [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) logic is computed asynchronously. This is very useful for responses that require more intensive computation and allows a [`Router`](xref:Bonsai.ZeroMQ.Router) to deal with frequent incoming [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) requests efficiently. 

> [!Note]
> Imagine, for example, that our Dealer sockets were sending video snippets to a Router server that is tasked with doing some processing of the video and returning the results back to the Dealers. If the responses were not computed in an asynchronous manner we would start to incur a bottleneck on the router if there were many connected Dealers or frequent Dealer requests.

> Running this workflow, you should see a 'bounceback' where any [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) client that sends a message receives a reply from the [`Router`](xref:Bonsai.ZeroMQ.Router) server. However, in order to address these messages to specific other clients we need to take a slightly different approach. 

Reviewer: ```suggestion
Running this workflow, you should see a 'bounceback' where any [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) client that sends a message receives a reply from the [`Router`](xref:Bonsai.ZeroMQ.Router) server. However, in order to address these messages to specific other clients we need to take a slightly different approach. Instead of using the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator we are going to build messages ourselves and pass them directly into the [`Router`](xref:Bonsai.ZeroMQ.Router). 
```

Reviewer: I am suggesting this and some of the following edits to break up the explanations and attach them more directly to each step as this section is still a little hard to follow.

- Delete the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) and [`ConvertToString`](xref:Bonsai.ZeroMQ.ConvertToString) branches.
- Create a [`BehaviorSubject`](xref:Bonsai.Reactive.BehaviorSubject) source with a `NetMQMessage` output type and name it 'RouterMessages' (right-click on an operator with a `NetMQMessage` output type >> CreateSource >> BehaviorSubject). Connect it as an input to the [`Router`](xref:Bonsai.ZeroMQ.Router).
- Delete the `Request.First` and `Buffer` outputs from [`Router`](xref:Bonsai.ZeroMQ.Router). Right-click on the [`Router`](xref:Bonsai.ZeroMQ.Router) and expose the `NetMQMessage` >> `First` >> `Buffer` output.

> - Add a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) called `BouceBack`. Connect [`Router`](xref:Bonsai.ZeroMQ.Router) to `BounceBack`.

Reviewer: ```suggestion
- Add a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) called `BounceBack`. Connect [`Router`](xref:Bonsai.ZeroMQ.Router) to `BounceBack`.

:::workflow
![Server message multicast](~/workflows/server-message-multicast.bonsai)
:::

Since the output type of the [`Router`](xref:Bonsai.ZeroMQ.Router) operator has changed from a  `ResponseContext` to a `NetMQMessage` we made some modifications to how we process the stream.  Inside the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator we will construct messages by splitting the `NetMQMessage` into its component `NetMQFrame` parts, extracting the relevant frames and merging them together.
```

Reviewer: This seems like a good area to separate the steps as they were getting long. I would also move the high level overview of the workflow here (server-message-multicast). I am not sure if the suggested change to the explanation is right though.

Author: Agreed, I slightly modified your suggestion to remove an extra whitespace and clarify briefly why the output type of the operator changes.

- Inside `BounceBack`, expose the `First` property of the `Source1` output.
- On a separate branch, add an [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator with an index 'Value' of 1. Connect `Source1` as its input.

> - On a further separate branch, add a [`String`](xref:Bonsai.Expressions.StringProperty) operator with a 'Value' of 'ServerResponse'. Connect this to a [`ConvertToFrame`](xref:Bonsai.ZeroMQ.ConvertToFrame) operator. Connect `Source1` as an input to the [`String`](xref:Bonsai.Expressions.StringProperty).

Reviewer: ```suggestion
- Inside `BounceBack`, expose the `First` property of the `Source1` output, which will give us the address of the `NetMQMessage`.
- On a separate branch, add an [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator with an index 'Value' of 1. Connect `Source1` as its input. This allows us to grab the middle empty delimiter frame.
- On a further separate branch, add a [`String`](xref:Bonsai.Expressions.StringProperty) operator with a 'Value' of 'ServerResponse'. Connect this to a [`ConvertToFrame`](xref:Bonsai.ZeroMQ.ConvertToFrame) operator. Connect `Source1` as an input to the [`String`](xref:Bonsai.Expressions.StringProperty). This will serve as the message content.
```

- Use [`Merge`](xref:Bonsai.Reactive.Merge) to combine the outputs of these 3 branches (`Source1.First`, `Index`, `ConvertToFrame`).
- Convert the output to a `NetMQMessage` by connecting a [`Take`](xref:Bonsai.Reactive.Take) operator with a 'Count' property of 3 followed by a [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) operator. 
- Finally, connect the output of [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) to a [`MulticastSubject`](xref:Bonsai.Expressions.MulticastSubject) targeting `RouterMessages`. Connect the [`MulticastSubject`](xref:Bonsai.Expressions.MulticastSubject) output to the `WorkflowOutput`.

:::workflow

> ![Server message multicast](~/workflows/server-message-multicast.bonsai)

Reviewer: ```suggestion
:::workflow
![BounceBack](~/workflows/server-bounceback.bonsai)
```

Reviewer: Suggested change to replace the high level overview workflow with one that provides an `inside peek` at the `BounceBack` operator as some people might mess up on those steps. Similar to what was done for the `Miss` and `Reward` sections of the [State Machines](https://bonsai-rx.org/docs/tutorials/state-machines.html) tutorial. The `Bounceback` workflow does not currently exist and will have to be created. I do realize though that many of the later steps also include logic that is hidden in a `SelectMany` or `Condition` operator and that including an inside peek every time might make the article too long and unwieldy so I am not sure if you want to consider including this.

:::

We had to change quite a few things to modify this workflow so let's step through the general logic. The first thing to note is that since we are avoiding the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator in this implementation we need to pass messages directly into the [`Router`](xref:Bonsai.ZeroMQ.Router). To do this we generate a [`BehaviorSubject`](xref:Bonsai.Reactive.BehaviorSubject) source with a `NetMQMessage` output type and connect it to the [`Router`](xref:Bonsai.ZeroMQ.Response). This will change the output type of the [`Router`](xref:Bonsai.ZeroMQ.Router) operator from a `ResponseContext` to a `NetMQMessage` so we need to make some modifications to how we process the stream.

> We want the [`Router`](xref:Bonsai.ZeroMQ.Router) to generate a reply message every time it receives a request from a [`Dealer`](xref:Bonsai.ZeroMQ.Dealer). Since we are now building this message ourselves instead of using [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse), we branch off the [`Router`](xref:Bonsai.ZeroMQ.Router) with a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator. Inside, we split the `NetMQMessage` into its component `NetMQFrame` parts, taking the `First` frame for the address, using [`Index`](xref:Bonsai.Expressions.IndexBuilder) to grab the middle empty delimiter frame and creating a new `String` which we convert to a `NetMQFrame` for the message content. We [`Merge`](xref:Bonsai.Reactive.Merge) these component frames back together and use a [`Take`](xref:Bonsai.Reactive.Take) operator (with count = 3) followed by [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage). The [`Take`](xref:Bonsai.Reactive.Take) operator is particularly important here as 1) [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) will only complete the message once the observable stream is completed and 2) We need to close the observable anyway to complete the [`SelectMany`](xref:Bonsai.Reactive.SelectMany). Finally, we use a [`MulticastSubject`](xref:Bonsai.Expressions.MulticastSubject) to send our completed message to the [`Router`](xref:Bonsai.ZeroMQ.Router).

Reviewer: ```suggestion
The [`Take`](xref:Bonsai.Reactive.Take) operator is particularly important here as 1) [`ToMessage`](xref:Bonsai.ZeroMQ.ToMessage) will only complete the message once the observable stream is completed and 2) We need to close the observable anyway to complete the [`SelectMany`](xref:Bonsai.Reactive.SelectMany). We use a [`MulticastSubject`](xref:Bonsai.Expressions.MulticastSubject) to send our completed message to the [`Router`](xref:Bonsai.ZeroMQ.Router).
```

Reviewer: Most of this paragraph was duplicated in the step breakout, so I have moved the explanations up top and closer to where the steps are.

> ---
> uid: client-server
> title: "Client-Server"
> ---
> 
> Client-server tutorial
> ======================

Reviewer 2: ```suggestion
Client-Server
=============
```

If we set the main article header to the desired article name, then we do not need to specify the article `title` and `uid` as it should be inferred correctly from the title and file name, respectively.

I guess we did use this explicit scheme in the first version of the articles, so we could keep it now for consistency and change in a separate PR, but maybe we can consider changing the main header text?

The Bonsai.ZeroMQ package allows us to harness the powerful [ZeroMQ](https://zeromq.org/) library to build networked applications in Bonsai. Applications could include:
- Interfacing with remote experimental rigs via network messages
- Performing distributed work across pools of machines (e.g. for computationally expensive deep-learning inference) 
- Streaming video data between clients across a network

> - **Real-time interaction between clients in a multiplayer game**

Reviewer 2: This bit seems like a summary of the ZeroMQ library. Do we need to include it as part of the article? Maybe some of these examples could be interesting to move to the intro itself?

In this article, we will use Bonsai.ZeroMQ to explore this final example and build a basic client-server architecture similar to one that might be used in a multiplayer game.

## Network design

> The basic network architecture we want to achieve will be composed of a number of clients sending their state to a server, which then updates the other connected clients with that clients’ state. This is comparable to a multiplayer game in which client players move through the game world and must synchronise that movement via a dedicated server so that all players see each other in their ‘true’ position in the world.

Reviewer 2: ```suggestion
The basic network architecture we want to achieve will be composed of a number of clients sending their state to a server, which then updates the other connected clients with that clients’ state. This is comparable to a multiplayer game in which players connect to a game server, move through a shared game world, and must synchronise that movement via a dedicated server so that all players see each other in their ‘true’ position in the world.
```

```mermaid
sequenceDiagram
    actor Client1
    actor Client2
    actor Client3
    participant Server
    Client1->>Server: MOVE
    Server->>Client2: SYNC
    Server->>Client3: SYNC
```

> An important requirement to point out here is that our server should be choosy about which clients it broadcasts information to. If client 1 updates the server with its current state, that information needs to be sent to all other connected clients, but there is no need to send it back to client 1 as it already knows its own state and this feedback message would be redundant.

Reviewer 2: ```suggestion
An important requirement to point out here is that in this case we want our server to be choosy about which clients it broadcasts information to. If client 1 updates the server with its current state, that information should be sent to all other connected clients, but there is no need to send it back to client 1 as it already knows its own state and this feedback message would be redundant.
```

Minor tweak to wording to avoid implying this is how all servers need to operate.

ZeroMQ provides a number of socket types that could be used to achieve something approaching this architecture. The @router-dealer socket pair acting as Server / Client has a couple of advantages for this design: 

> - Routers assign a unique address for each connected client allowing clients in turn to be addressed individually

Reviewer 2: ```suggestion
- Routers assign a unique address to each connected client, allowing clients in turn to be addressed individually
```

- Messages can be passed between Router / Dealer sockets without the requirement that a reply is received before the next message is sent, as is the case with the Request / Response socket pair.

## Basic client
To begin with, we’ll create a simple client that sends basic messages on a network. In a new Bonsai project:

> - Add a [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator. 

Reviewer 2: ```suggestion
- Insert a [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator. 
```

For consistency with other tutorials and with the name of the command in editor context menus, we tend to prefer "Insert" instead of "Add".

> - In the `ConnectionString` property, set `Address`: localhost:5557, `Action`: Connect, `Protocol`: TCP.

Reviewer 2: ```suggestion
- Set the following values in the `ConnectionString` property:
  - `Address`: localhost:5557
  - `Action`: Connect
  - `Protocol`: TCP
```

Was experimenting with different layouts to present the sub-properties, and found that the sub-bulletlist seemed to capture closest what one sees in the property grid, but happy to hear thoughts on this.

> [!Note]

> > In Bonsai.ZeroMQ, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) can have two functions based on its inputs. On its own, as above, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator creates a Dealer socket that listens for messages on the specified network. With the properties specified, we are asking our [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) to listen for messages on the local machine on port 5557 using the TCP protocol. We use the ‘Connect’ argument for the `SocketConnection` property to tell the dealer that it will connect to a static part of the network with a known IP address, in this case the server which we will implement later.

Reviewer 2: ```suggestion
> In `Bonsai.ZeroMQ`, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator can perform two functions based on its inputs. Placed on its own, as above, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator creates a socket which listens for messages on the specified network. With the properties specified, we are asking our [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) to listen for messages on the local machine on port 5557 using the TCP protocol. We use the ‘Connect’ argument on the `Action` property to tell the dealer that it will connect to a node in the network having a known IP address, in this case the server which we will implement later.
```

Minor tweaks and styling recommendations. I was thinking these notes might actually be useful to include in extended remarks on the `Dealer` operator documentation itself. Then we could refer to those remarks in this and any future examples.

If we add inputs to the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer), it will act as both a sender and receiver of messages on the network. 

> - Before the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator add a [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown). 
> - Add a [`String`](xref:Bonsai.Expressions.StringProperty) operator in sequence as input to the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer).

Reviewer 2: ```suggestion
- Insert a [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) operator before the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer).
- Insert a [`String`](xref:Bonsai.Expressions.StringProperty) operator in the chain, as input to the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer).
```

In general I try to start the sentences in tutorials with the action verb of the specific step, so that looking at the left-hand side of all bullet points you can quickly skim what kind of action it will require.

This is more of a personal style consistency point at the moment, so feel free to take or drop it.

I also tried to think of an alternative to the word "sequence" that we could use, to avoid confusion with observable sequences, and I came up with the word "chain", as in "chaining operators", but maybe there are better terms.

:::workflow
![Basic Dealer input](~/workflows/dealer-basic-input.bonsai)
:::

> - In the operator properties, set the [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) `Filter` to the ‘1’ key.
> - Set the [`String`](xref:Bonsai.Expressions.StringProperty) `Value` to ‘Client1’. 

Reviewer 2: ```suggestion
- Set the `Filter` property of the [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) operator to the `1` key.
- Set the `Value` property of the [`String`](xref:Bonsai.Expressions.StringProperty) operator to `Client1`.
```

For consistency with other tutorials we keep the object of the verb close for clarity, and qualify as needed, so the main action in this case is "Set the `Value` property".

I was also checking in other tutorials, and we only use the xref link to the operator on the very first step we reference it (usually the insert step). However, I agree it might be more consistent, if a bit tedious, to just keep adding the hyperlinks.

The last thing I was curious about is how we should refer to string literals in tutorials. Usually for numeric values we just write the number, but in this case it is a string, so it feels somewhat correct programming-wise to escape it with quotes. However, in the editor property grid you actually never see quotes anywhere and just write the string, so I'm wondering whether this could confuse people who might try to be literal and type the entire thing as-is in the property grid entry only to get an error.

I was looking into other similar tutorials (e.g. [Networking](https://bonsai-rx.org/docs/tutorials/networking.html)) and I noticed we were using mostly verbatim literals for strings. Not sure how clear that is since we also use verbatim for property names, but I guess there is no rule in code markdown that verbatim literals should be used to emphasize only one type of code, so went for that in the recommendations. II guess we could also think of using **bold** or _italic_, happy to hear thoughts though.

Reviewer: Just wanted to add a comment here that it has been a bit inconsistent with the other tutorials, I do prefer adding the xref for all operator references as it also helps distinguish the operators from the properties. Making a note to raise a PR to align the formatting on the main bonsai docs. 

As for the tediousness of it I was working on a script to automatically xref the markdown files, maybe once I polish and finish it I can share it. Also taking note of all the other tips in this review!

> If we run the Bonsai project now, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) will continue listening for incoming messages on the network, but every time the ‘1’ key is pressed a message containing the string ‘Client1’ will be sent from the socket.

Reviewer 2: ```suggestion
If you run the workflow, the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) will continue listening for incoming messages on the network, but every time the `1` key is pressed a message containing the string `Client1` will be sent from the socket.
```

This is again a minor observation on style, but after several iterations on the tutorials, we ended up converging on writing in a more conversational style for tutorials, preferring the use of the "you" pronoun over "we". We can still mix in "we" as it feels natural.

Also sometimes when introducing a step as an investigation, we tend to start by asking them to run the workflow, and then mention what they should notice.

> - Copy and paste this client structure a couple of times and change the [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) and [`String`](xref:Bonsai.Expressions.StringProperty) properties accordingly on each (2, ‘Client2’; 3, ‘Client3’) so that we have 3 total clients that send messages according to different key presses.

Reviewer 2: ```suggestion
- Copy and paste this client structure a couple of times and change the [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) and [`String`](xref:Bonsai.Expressions.StringProperty) properties accordingly on each (2, `Client2`; 3, `Client3`) so that we have a total of 3 clients which send messages according to different key presses.
```

:::workflow
![Multiple clients](~/workflows/multiple-clients.bonsai)
:::

> [!Note]

> > For the purposes of this article we are creating all of our clients and our server on the same Bonsai project and same machine for ease of demonstration. In a working example, each client and server could be running in separate Bonsai instances on different machines on a network. In this case, localhost would be replaced with the server machine’s IP address.

Reviewer 2: ```suggestion
> For the purposes of this article we are creating all of our clients and our server on the same workflow running on the same machine for ease of demonstration. In a real-world case, each client and server could be running in separate workflows installed on different machines in a network. In this case, `localhost` would be replaced with the server machine’s IP address.
```

## Basic server
Now that we have our client pool set up and sending messages, let’s implement a server to listen for those messages. 
- Add a [`Router`](xref:Bonsai.ZeroMQ.Router) operator to the project and set its properties to match the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) sockets we already added so that it is running on the same network. 

> - As the [`Router`](xref:Bonsai.ZeroMQ.Router) is acting as server and will be the ‘static’ part of the network, set its `Action` to ‘Bind’.

Reviewer 2: It might be worth here to specify exactly the value of the properties to set, even if they are almost identical (except for the `Bind` value). This way people can check their understanding of the parametrization of these socket pairs, i.e. where they are the same, and where they differ.

> As with the [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) operator, a [`Router`](xref:Bonsai.ZeroMQ.Router) operator without any input will simply listen for messages on the network and not send anything in return. If we run the project now and monitor the output of the [`Router`](xref:Bonsai.ZeroMQ.Router) operator, we'll see that each time the client sends a message triggered by its associated key press we get a `ResponseContext` produced at the [`Router`](xref:Bonsai.ZeroMQ.Router). Expanding the output the [`Router`](xref:Bonsai.ZeroMQ.Router), we can see it contains a `NetMQMessage`. We [expect](https://netmq.readthedocs.io/en/latest/router-dealer/) this message to be composed of 3 frames: an address (in this case the address of the client that sent the message), an empty delimiter frame and the message content. 

Reviewer 2: Reading this now, I wonder whether it might be better to swap the `Router` and the `Dealer` steps. That way you would naturally start with a source operator, rather than a combinator, and you would also have an immediate feedback to test with as soon as you create your first `Dealer`.

> - Expose the `Buffer` `byte[]` of the `First` frame. 

Reviewer 2: ```suggestion
- Insert a `MemberSelector` operator and select the `First` > `Buffer` member.
```

The verb "Expose" might be too ambiguous, so maybe it is better to be explicit and either ask specifically to insert a `MemberSelector` or converge on some other standard way to refer to this action. Sometimes in the past we have used right-clicking, but maybe we could come up with a general term for this.

- Add an [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator the first frame buffer and set its `Value` to 1 to access the unique address ID. 

> - Add a [`ConvertToString`](xref:Bonsai.ZeroMQ.ConvertToString) to the `Last` frame. 

Reviewer 2: If the above recommendations are adopted, replace with `Insert`.

So far our network is rather one-sided. We can send client messages to the server which can in turn receive and parse them, but currently nothing is relayed back to the clients. The first goal for server feedback is that any time a client message is received, the server sends this message back to all connected clients. To do this, we first need a way of keeping track of all active clients. 

- Add a [`Zip`](xref:Bonsai.Reactive.Zip) operator to the [`Index`](xref:Bonsai.Expressions.IndexBuilder) operator and connect the `byte[]` `Buffer` as the second input.

:::workflow
![Address key-value pair](~/workflows/address-kvp.bonsai)
:::

Every time the [`Router`](xref:Bonsai.ZeroMQ.Router) receives a message, the [`Zip`](xref:Bonsai.Reactive.Zip) will create a `Tuple` that can be thought of as a key-value pair, with the unique `byte` address of the client as the key, and the full `byte[]` address used by ZeroMQ for routing as the value. 

- Add a [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) operator after the [`Zip`](xref:Bonsai.Reactive.Zip) and set the `KeySelector` property to the `byte` value (`Item1`).

:::workflow
![Unique key-value pair](~/workflows/unique-kvp.bonsai)
:::

> The [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) operator filters the output of [`Zip`](xref:Bonsai.Reactive.Zip) according to the unique `byte` value and produces a sequence containing only the distinct – or ‘new’ – values produced by [`Zip`](xref:Bonsai.Reactive.Zip). The output of [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) will therefore effectively be a sequence of unique client addresses corresponding to each connected client. We also need to store these unique values and make them available to other parts of the Bonsai workflow. 

Reviewer 2: ```suggestion
The [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) operator filters the output of [`Zip`](xref:Bonsai.Reactive.Zip) according to each `byte` value and produces a sequence containing only tuples with the first unique byte values produced by [`Zip`](xref:Bonsai.Reactive.Zip). The output of [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) will therefore effectively be a sequence of unique client addresses corresponding to each connected client. Now we need to store these unique values and make them available to other parts of the Bonsai workflow. 
```

- Add a [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) operator after [`DistinctBy`](xref:Bonsai.Reactive.DistinctBy) and name it ‘ClientAddresses’. 

:::workflow
![Address ReplaySubject](~/workflows/address-replay-subject.bonsai)
:::

> A [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) has the useful feature that it stores its input sequence and replays those values to any current or future subscribers. The effect in this case is that anything that subscribes to `ClientAddresses` will receive all the unique client addresses encountered by the server so far.

Reviewer 2: ```suggestion
By using a [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) we can store the sequence of unique client addresses and replay it to any current or future subscribers. This subscription is dynamic, so any operator subscribing to `ClientAddresses` will immediately receive all the unique client addresses encountered by the server so far, and also any future ones as soon as they are seen the first time by the server.
```

## Server --> client communication

> Eventually, we will use these unique client addresses to route server messages back to specific client. For now, we'll implement a more basic approach where the server just sends messages back to the client that originally sent them. The Bonsai ZeroMQ library provides a convenient operator for this task in the form of [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse). 

Reviewer 2: ```suggestion
Eventually, we will use these unique client addresses to route server messages back to a specific client. For now, we will implement a more basic approach where the server just sends messages back to the client that originally sent them. `Bonsai.ZeroMQ` provides a convenient [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator to make this task much easier. 
```

- Add a [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator after the [`Router`](xref:Bonsai.ZeroMQ.Router) in a separate branch.

> - Inside (double-click on [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse)) add a [`String`](xref:Bonsai.Expressions.StringProperty) operator with a generic response value like `ServerResponse` after the `Source` operator. 

Reviewer 2: ```suggestion
- Double-click on the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) node to open up its internal workflow.
- Insert a [`String`](xref:Bonsai.Expressions.StringProperty) operator after the workflow `Source`, with a generic response value like `ServerResponse`. 
```

From other tutorials, we tend to describe the step of opening up a nested workflow separately from what we do inside.

:::workflow
![Basic server response](~/workflows/server-basic-response.bonsai)
:::

> The [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator has a couple of interesting properties which may not be immediately obvious from this simple example. First, this operator always transmits its response back to the ZeroMQ socket that initiated the request (in this case one of our [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) clients) and we therefore do not need to specify an address in its processing logic. Second, the internal flow of [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) logic is computed asynchronously. This is very useful for responses that require more intensive computation and allows a [`Router`](xref:Bonsai.ZeroMQ.Router) to deal with frequent incoming [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) requests efficiently. 

Reviewer 2: I wonder if an analogy or connection with `SelectMany` should be made here, explaining that it is an operator that also runs in parallel for each request, but with an additional side-effect of sending the response back to the client.

> [!Note]
> Imagine, for example, that our Dealer sockets were sending video snippets to a Router server that is tasked with doing some processing of the video and returning the results back to the Dealers. If the responses were not computed in an asynchronous manner we would start to incur a bottleneck on the router if there were many connected Dealers or frequent Dealer requests.

> Running this workflow, you should see a 'bounceback' where any [`Dealer`](xref:Bonsai.ZeroMQ.Dealer) client that sends a message receives a reply from the [`Router`](xref:Bonsai.ZeroMQ.Router) server. However, in order to address these messages to specific other clients we need to take a slightly different approach. Instead of using the [`SendResponse`](xref:Bonsai.ZeroMQ.SendResponse) operator we are going to build messages ourselves and pass them directly into the [`Router`](xref:Bonsai.ZeroMQ.Router). 

Reviewer 2: I guess adding the previous suggestion could help explain why we are going the route of `SelectMany` in this case.

:::workflow
![SelectMany detour](~/workflows/select-many-detour.bonsai)
:::

- Run the project and inspect the output of the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator. 
- If no client messages are triggered and we press ‘A’ to trigger the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) nothing will be returned. 
- If we trigger a single client and press ‘A’ again [`SelectMany`](xref:Bonsai.Reactive.SelectMany) gives us the address of that client. 
- If we trigger a second client and press ‘A’ we get the addresses of these first two clients in sequence, and so on if we add the third client. 

Whenever we press ‘A’ we get a sequence of all the connected client addresses. Every time we trigger [`SelectMany`](xref:Bonsai.Reactive.SelectMany) with a [`KeyDown`](xref:Bonsai.Windows.Input.KeyDown) we generate a new sequence that immediately subscribes to `ClientAddresses`, a [`ReplaySubject`](xref:Bonsai.Reactive.ReplaySubject) which replays all our unique client addresses into the sequence. 

> We could keep initiating these new sequences by continually pressing ‘A’ and if a new client address were to be added then all these sequences would report the new address (you can test this by connecting the [`SusbcribeSubject`](xref:Bonsai.Expressions.SubscribeSubject) directly to the workflow output and deleting [`KeyUp`](xref:Bonsai.Windows.Input.KeyUp) and [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil)). Instead, we want to complete each new sequence once it has given us all the client addresses so we use an arbitrary event (releasing the key that initiated the sequence) to trigger [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil) and close the sequence. The overall effect is something similar to a loop that iterates over all client addresses every time we request it with a key press. This is the general structure of what we want to achieve next in our server logic to broadcast messages back to all connected clients.

Reviewer 2: It feels like the `SelectMany` detour should point to an article on the core Bonsai documentation which explains how the operator works. I understand how people may be confused by this operator coming in new to Bonsai, but it is such a prevalent operator that this issue will come up in almost every mildly complex tutorial we write about almost anything.

It might be better to point all of these `SelectMany` references to the same place, and focus on improving that central hub, as it is one of the core concepts in Bonsai, so that it is reachable from almost any entry point from which you may want to learn the language.

It is probably still useful, however, to explain how `SelectMany` is being applied in this case, i.e. for what purpose, and explain the effects of using it that way so people can relate to their understanding of the general operator.

## All client broadcast
To apply the logic of the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) example to server broadcast, we need something to trigger the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) sequence creation, and something to trigger termination. We already have a trigger for sequence creation in the output of the [`Router`](xref:Bonsai.ZeroMQ.Router) since we want to run our [`SelectMany`](xref:Bonsai.Reactive.SelectMany) sequence every time a client message is received. For our sequence temination trigger, we want something that is guaranteed to fire after the server receives a client message, but before the next message is received so that our [`SelectMany`](xref:Bonsai.Reactive.SelectMany) sequence for each message responds only to that particular message. A simple solution is therefore to use the arrival of the next message as our sequence termination trigger.

- Add a [`Skip`](xref:Bonsai.Reactive.Skip) operator after the [`Router`](xref:Bonsai.ZeroMQ.Router) in a separate branch and connect this to a [`PublishSubject`](xref:Bonsai.Reactive.PublishSubject). 

> - Set the [`Skip`](xref:Bonsai.Reactive.Skip) operator's `Count` property is set to 1, and name the [`PublishSubject`](xref:Bonsai.Reactive.PublishSubject) 'NextMessage'.

Reviewer 2: ```suggestion
- Set the [`Skip`](xref:Bonsai.Reactive.Skip) operator's `Count` property is set to 1, and name the [`PublishSubject`](xref:Bonsai.Reactive.PublishSubject) `NextMessage`.
```

Using verbatim literals for string property values.

- Add a [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator after the [`Router`](xref:Bonsai.ZeroMQ.Router) in a separate branch and name it ‘SelectAllClients’.
- Inside the [`SelectMany`](xref:Bonsai.Reactive.SelectMany) operator, create 2 [`SubscribeSubject`](xref:Bonsai.Expressions.SubscribeSubject) operators and link them to the `ClientAddresses` and `NextMessage` subjects. 
- Connect the `ClientAddresses` subscription to the workflow output via a [`TakeUntil`](xref:Bonsai.Reactive.TakeUntil) operator and use `NextMessage` as the second input. 

Now, our `SelectAllClients` will produce a sequence of all unique client addresses every time the server receives a message. Connect the output of `SelectAllClients` to a [`WithLatestFrom`](xref:Bonsai.Reactive.WithLatestFrom) with the [`Router`](xref:Bonsai.ZeroMQ.Router) as its second input. In this context [`WithLatestFrom`](xref:Bonsai.Reactive.WithLatestFrom) combines each client address from `SelectAllClients` with the most recent received message. The result is that when a message is received from the client, we produce several copies of the message 'addressed' to each connected client.

:::workflow

> ![Select all clients and package message](~/workflows/format-select-all-clients.bonsai)

Reviewer 2: Seems like this workflow has a disabled or invalid `BounceBack` node which perhaps should be removed to avoid confusion if not needed?
